### PR TITLE
feature: poetry show json output format

### PR DESCRIPTION
**Pull request by Aearsears**
# Pull Request Check List

Resolves: #10451

Adds a `--output` option to the `poetry show` command for `text` and `json` outputs.

`poetry self show` is not updated and there is some test duplication in `test_show.py` but this PR doesn't update them as I wanted to keep changes as contained as possible.

Duplicated test coverage:
- `test_show_non_dev_with_basic_installed_packages` captured in `test_show_basic_with_group_options --without test`
- `test_show_with_group_only` captured in `test_show_basic_with_group_options --only {MAIN_GROUP}`
- `test_show_with_optional_group` captured in `test_show_basic_with_group_options --with time`

I will open another PR for the duplicated tests.

- [x] Added **tests** for changed code.
- [x] Updated **documentation** for changed code.

## Summary by Sourcery

Implement an --output flag for poetry show that enables JSON output alongside the existing text format, add validation for invalid or conflicting options, serialize package data to JSON in appropriate code paths, update tests accordingly, and document the new flag.

New Features:
- Add an --output option to the poetry show command to select between text and JSON output
- Support JSON formatting for both single-package and multi-package show invocations

Enhancements:
- Validate the output format and disallow combining JSON output with the --tree option
- Serialize package information to JSON in both single and list display modes

Documentation:
- Document the --output option in the CLI reference and note supported formats and constraints

Tests:
- Extend existing show command tests to verify JSON output across all options and edge cases
- Add tests for invalid output formats and tree/JSON incompatibility errors

## Summary by Sourcery

Add an `--output` flag to the `poetry show` command to support JSON output alongside text, with validation for invalid formats and conflicting `--tree` usage

New Features:
- Add `--output` option for the `poetry show` command to choose between text and JSON formats
- Enable JSON output for both single-package and multi-package displays, including group filters, top-level, outdated, latest, and reverse-dependency views

Enhancements:
- Validate the `--output` argument and emit errors for unsupported formats or invalid `--tree` combinations
- Refactor the show command to handle JSON serialization early and streamline display logic

Documentation:
- Document the new `--output` option in the CLI reference with supported formats and constraints

Tests:
- Parameterize existing show tests to cover both text and JSON output across all modes
- Add tests for invalid `--output` values and JSON/`--tree` incompatibility errors

### CODE REVIEW

**Comment by Aearsears** [comment]
@sourcery-ai review

**Comment by Aearsears** [comment]
question for the reviewer on the ai bot's comment: `In show.py you import json inside multiple blocks; moving it to the module scope would remove duplicate imports and improve readability.`

On [line 257](https://github.com/python-poetry/poetry/blob/73ad0e33bd3416460ff4ab1bb3a7ca2bdfc09903/src/poetry/console/commands/show.py#L257) and then on [line 638](https://github.com/python-poetry/poetry/blob/73ad0e33bd3416460ff4ab1bb3a7ca2bdfc09903/src/poetry/console/commands/show.py#L638), it imports `from cleo.io.null_io import NullIO` twice. for the `json` package is it fine to import it on the module scope? will this affect poetry's performance?

**Comment by radoering** [comment]
> for the `json` package is it fine to import it on the module scope? will this affect poetry's performance?

It should be fine. Unfortunately, there is no documentation about which imports are expensive and should not be done at top level. The only way to tell for sure is to do some measurements: Run `time poetry show -h` (which imports the command but does not actually execute it) multiple times with local/global imports and check if there are significant differences. `python -X importtime` might also be useful to assess if an import is expensive.

File: src/poetry/console/commands/show.py [hunk 1]
```diff
@@ -306,23 +351,69 @@ def _display_packages_information(
...
+                    latest = latest_packages[locked.pretty_name]
+                    package["latest_version"] = get_package_version_display_string(
+                        latest, root=self.poetry.file.path.parent
+                    )
+
+                if show_why:
+                    required_by = reverse_deps(locked, locked_repository)
+                    if required_by:
+                        content = ",".join(required_by.keys())
+                        package["required_by"] = content
```
[the remark this answers is not in the document]
Comment (Aearsears) [reply src/poetry/console/commands/show.py]: should not be an issue with pep package naming guidelines

**Comment by Aearsears** [comment]
for single package its a single json object
```json
{
"name": "requests", "version": "2.32.3", "description": "Python HTTP for Humans.", 
"dependencies": {"certifi": ">=2017.4.17", "charset-normalizer": ">=2,<4", "idna": ">=2.5,<4", "urllib3": ">=1.21.1,<3"}, 
"required_by": {"cachecontrol": ">=2.16.0", "requests-toolbelt": ">=2.0.1,<3.0.0"}
}
```

for all the packages it will output a list of objects like
```json
[
{"name": "virtualenv", "installed_status": "installed", "version": "20.30.0", "latest_version": "20.32.0", "description": "Virtual Python Environment builder"}, 
{"name": "zstandard", "installed_status": "installed", "version": "0.23.0", "latest_version": "0.23.0", "description": "Zstandard bindings for Python"}
]
```

File: src/poetry/console/commands/show.py [hunk 2]
```diff
@@ -306,23 +350,67 @@ def _display_packages_information(
...
```
[the remark this answers is not in the document]
Comment (Aearsears) [reply src/poetry/console/commands/show.py]: should not be an issue with pep package naming guidelines

File: src/poetry/console/commands/show.py [hunk 3]
```diff
@@ -37,6 +38,9 @@ def reverse_deps(pkg: Package, repo: Repository) -> dict[str, str]:
     return required_by
 
 
+OUTPUT_FORMATS = {"text", "json"}
```
Comment (colindean) [inline src/poetry/console/commands/show.py]: **nitpick:** I would guide toward an Enum for this. I find them nicer to work with in the short and long term.

**Review by colindean — COMMENTED** [review]
Thank you for undertaking this!

I add one little nit, a personal preference for Enums over lists of strings for enumerations of valid options. I always gently guide folks toward Enums in these situations.

File: src/poetry/console/commands/show.py (hunk 3 above)
Comment (radoering) [reply src/poetry/console/commands/show.py]: This would probably have to be a `StrEnum`, since it must be a string? However, `StrEnum` has just been introduced in Python 3.11 and we still have to support Python 3.9 and 3.10. Thus, we would either have to define our own `StrEnum` or keep it as is and convert it to `StrEnum` once we drop support for Python 3.9 and 3.10, i.e. their EOL.

File: src/poetry/console/commands/show.py [hunk 4]
```diff
@@ -71,6 +75,13 @@ class ShowCommand(GroupCommand, EnvCommand):
             None,
             "Do not truncate the output based on the terminal width.",
         ),
+        option(
+            "output",
```
Comment (radoering) [inline src/poetry/console/commands/show.py]: `"format"` might be a better name for the option. In [poetry-plugin-export](https://github.com/python-poetry/poetry-plugin-export?tab=readme-ov-file#available-options), `--format` is the output format and `--output` is the name of the output file.

File: src/poetry/console/commands/show.py [hunk 5]
```diff
@@ -306,23 +350,67 @@ def _display_packages_information(
...
+                if show_latest:
+                    latest = latest_packages[locked.pretty_name]
+                    package["latest_version"] = get_package_version_display_string(
+                        latest, root=self.poetry.file.path.parent
+                    )
+
+                if show_why:
+                    required_by = reverse_deps(locked, locked_repository)
+                    if required_by:
+                        content = ",".join(required_by.keys())
```
Comment (radoering) [inline src/poetry/console/commands/show.py]: Wouldn't it make more sense to make this a list instead of a string of comma separated values?

File: docs/cli.md [hunk 6]
```diff
@@ -1061,6 +1061,7 @@ required by
 * `--all (-a)`: Show all packages (even those not compatible with current system).
 * `--top-level (-T)`: Only show explicitly defined packages.
 * `--no-truncate`: Do not truncate the output based on the terminal width.
+* `--output`: Specify the output format. JSON cannot be combined with the `--tree` option. Default is text. Supported formats: text, json.
```
Comment (radoering) [inline docs/cli.md]:
```suggestion
* `--output`: Specify the output format (`text` or `json`). Default is `text`. `json` cannot be combined with the `--tree` option.
```

File: src/poetry/console/commands/show.py (hunk 5 above)
Comment (Aearsears) [reply src/poetry/console/commands/show.py]: yes, that's better and makes more sense. ive made it in a list of str and also added [test_show_why](https://github.com/python-poetry/poetry/pull/10487/files#diff-beb6cb30b284a820c167f2193f8031ba60a1354dfb8501874615e50137d39b15R2360) for this

File: src/poetry/console/commands/show.py (hunk 4 above)
Comment (Aearsears) [reply src/poetry/console/commands/show.py]: sure, ive changed it to `format` to better align with poetry's other tooling

File: src/poetry/console/commands/show.py (hunk 3 above)
Comment (Aearsears) [reply src/poetry/console/commands/show.py]: thanks radoering and colindean for your reviews. i read up on some docs. to support python 3.9 and 3.10, its now using the derived enumeration `class OutputFormats(str, Enum)`  for str equality comparisons. only python 3.12+ supports directly using `in`   to test if a value belongs to the enum so I have to convert it to a set first. once `StrEnum` is supported it will be simple to swap it in

File: docs/cli.md [hunk 7]
```diff
@@ -1061,6 +1061,7 @@ required by
 * `--all (-a)`: Show all packages (even those not compatible with current system).
 * `--top-level (-T)`: Only show explicitly defined packages.
 * `--no-truncate`: Do not truncate the output based on the terminal width.
+* `--output`: Specify the output format (`json` or `text`). Default is `text`. `json` cannot be combined with the `--tree` option.
```
Comment (radoering) [inline docs/cli.md]:
```suggestion
* `--format (-f)`: Specify the output format (`json` or `text`). Default is `text`. `json` cannot be combined with the `--tree` option.
```

Comment (Aearsears) [reply docs/cli.md]: thanks for catching this

**Review by radoering — APPROVED** [review]

**Comment by colindean** [comment]
@Aearsears thank you so very much for picking this up and running with it. This is great work and I can't wait to put it into use. 

@radoering thanks again for doing what you do! I recently onboarded a team of about 50 people to Poetry :-)
